### ISSUE TITLE
Regex transform

### ISSUE THREAD

**Issue by danielphan2003** [TrianguloY/UrlChecker#170]
**Is your feature request related to a problem? Please describe.**
When replacing a url with regex, there is some cases where you need to transform them. For example: example.com/HeY ➔ example.com/hey

**Describe a solution you've considered**
None.

**Describe alternatives you've considered**
None.

**Additional context**
None.

**Comment by TrianguloY** [comment]
Do you mean using a regexp transformation? like '/H(.)Y' -> '/h$1y'? The current patterns checker module does allow for a replacement, there are some included which uses it, and you can find more on the [wiki](https://github.com/TrianguloY/UrlChecker/wiki/Custom-patterns).

Or do you mean something else like changing the characters case (lower/upper)?

**Comment by danielphan2003** [comment]
Oh, I meant for the lower/upper case transformation.

**Comment by TrianguloY** [comment]
Ah, unfortunately java regexp doesn't support changing case. I could add a module with two buttons, or at least one to make the url lowercase. In any case, what is the reasoning for this feature? Aren't urls case-insensitive?

**Comment by danielphan2003** [comment]
tl;dr the scheme (e.g. `https` vs `HTTPS`) and the host of a URI are case-insensitive, but the rest (paths, queries etc.) are not, at least according to RFC 3986, RFC 2616 and RFC 7230

From an answer on https://stackoverflow.com/questions/15641694/are-uris-case-insensitive :
> [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986) states:
> > the scheme and host are case-insensitive and therefore should be normalized to lowercase. For example, the URI <HTTP://www.EXAMPLE.com/> is equivalent to <http://www.example.com/>. The other generic syntax components are assumed to be case-sensitive unless specifically defined otherwise by the scheme
>
> [RFC 2616](https://www.rfc-editor.org/rfc/rfc2616) defines the following comparison rule for the HTTP scheme:
> > When comparing two URIs to decide if they match or not, a client SHOULD use a case-sensitive octet-by-octet comparison of the entire URIs, with these exceptions:
>
> However, [RFC 7230](https://www.rfc-editor.org/rfc/rfc7230) locks it down further by stating
> > The scheme and host are case-insensitive and normally provided in lowercase; all other components are compared in a case-sensitive manner.
>
> Those rules typically apply to client side comparisons. There are no rules specifically geared for server side comparisons. Once a server breaks up a URI into its components, it should treat them according to the same rules, but I don't see that enforced in the RFCs. Some web servers, like Apache, do follow the rules. IIS doesn't, for compatibility with Windows' case-insensitive file system.

**Comment by TrianguloY** [comment]
So, for that reason, a module to change the url case is not a good idea, since it would break most of them. It could change the left part though, but I'm not sure if that's really useful in any case.

Since you mentioned regexp, is there any specific user case you know that may benefit from this feature? Maybe there is a shortener-like service that makes the long urls with different case and so you need to change it when unshortening it?

**Comment by TrianguloY** [comment]
That's... a very good reason indeed. According to the standard the domain part of an url is case insensitive, and some apps even translate it to lowercase. It's a shame that java regexp doesn't support case syntax, replacing the domain with a lowercase is a nice addition. I wonder if having it as a special rule in that module could be problematic. A new module just for that seems a bit too much, but with this new information maybe it's worth it.

I'm thinking of a module with two buttons:
- Change domain to lowercase: will notify if the domain part has uppercase, and pressing the button will make it lowercase. Probably also with an automatic setting toggle since the destination url doesn't change so it's safe.
- Change path case: if there is at least an uppercase letter, will make it all lowercase. If all are lowercase will make it uppercase. Not sure if this would be a good idea, but a regexp to specify which urls to auto-change can be useful...maybe? Paths in urls are case sensitive, so the destination does change and it's not safe.

**Comment by Bakr-Ali** [comment]
> I was actually searching for a way to change the uppercase characters to lowercase in the domain.
> 
> It's useful since I could send you a malicious link like https://googIe.com and it would be hard to notice there's an uppercase _i_ instead of a lowercase _l_. However, if the domain characters were all lowercase, it would be easier to determine that's a fake website.
> 
> It makes even more sense to add this if you consider the pattern checker, by default, checks for non-ascii characters to help recognize websites used for phishing, as explained in the app:
> 
> > - Warning when contains non-ascii characters like greek letters.
> > This can be used for phishing: googÍe.com vs google.com

If that's the only use case, then a similar warning for upper case letters can be easily added as a pattern in the Pattern checker module

```
  "Warning! Upper case letter in domain.": {
    "regex": "^.*?:\/\/[^\/?#]*[A-Z]"
  }
```

**Comment by TrianguloY** [comment]
> If that's the only use case, then a similar warning for upper case letters can be easily added as a pattern in the Pattern checker module
> 
> ```
>   "Warning! Upper case letter in domain.": {
>     "regex": "^.*?:\/\/[^\/?#]*[A-Z]"
>   }
> ```

That is actually a really good pattern. I think I'll add it to the app itself. Thanks!

**Comment by Bakr-Ali** [comment]
> That is actually a really good pattern. I think I'll add it to the app itself. Thanks!

I've managed to make a series of regex replacements to make it lowercase, too! Although it is _a bit_ verbose.

Try it with this link: `http://WWw.GgOoogIe.CCCcom/?ref=REFerrer&foo=BAR#tag`

```

  "Warning! Upper case in domain. Lowercase it?": {
    "regex": "^.*?:\/\/[^\/?#]*[A-Z]",
    "replacement": "makedomainlowercase$0"
  }
  "Domain Lowercase A": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([A])(.*)$",
    "replacement": "$1a$3",
    "automatic": "true"
  },
  "Domain Lowercase B": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([B])(.*)$",
    "replacement": "$1b$3",
    "automatic": "true"
  },
  "Domain Lowercase C": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([C])(.*)$",
    "replacement": "$1c$3",
    "automatic": "true"
  },
  "Domain Lowercase D": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([D])(.*)$",
    "replacement": "$1d$3",
    "automatic": "true"
  },
  "Domain Lowercase E": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([E])(.*)$",
    "replacement": "$1e$3",
    "automatic": "true"
  },
  "Domain Lowercase F": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([F])(.*)$",
    "replacement": "$1f$3",
    "automatic": "true"
  },
  "Domain Lowercase G": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([G])(.*)$",
    "replacement": "$1g$3",
    "automatic": "true"
  },
  "Domain Lowercase H": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([H])(.*)$",
    "replacement": "$1h$3",
    "automatic": "true"
  },
  "Domain Lowercase I": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([I])(.*)$",
    "replacement": "$1i$3",
    "automatic": "true"
  },
  "Domain Lowercase J": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([J])(.*)$",
    "replacement": "$1j$3",
    "automatic": "true"
  },
  "Domain Lowercase K": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([K])(.*)$",
    "replacement": "$1k$3",
    "automatic": "true"
  },
  "Domain Lowercase L": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([L])(.*)$",
    "replacement": "$1l$3",
    "automatic": "true"
  },
  "Domain Lowercase M": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([M])(.*)$",
    "replacement": "$1m$3",
    "automatic": "true"
  },
  "Domain Lowercase N": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([N])(.*)$",
    "replacement": "$1n$3",
    "automatic": "true"
  },
  "Domain Lowercase O": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([O])(.*)$",
    "replacement": "$1o$3",
    "automatic": "true"
  },
  "Domain Lowercase P": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([P])(.*)$",
    "replacement": "$1p$3",
    "automatic": "true"
  },
  "Domain Lowercase Q": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([Q])(.*)$",
    "replacement": "$1q$3",
    "automatic": "true"
  },
  "Domain Lowercase R": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([R])(.*)$",
    "replacement": "$1r$3",
    "automatic": "true"
  },
  "Domain Lowercase S": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([S])(.*)$",
    "replacement": "$1s$3",
    "automatic": "true"
  },
  "Domain Lowercase T": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([T])(.*)$",
    "replacement": "$1t$3",
    "automatic": "true"
  },
  "Domain Lowercase U": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([U])(.*)$",
    "replacement": "$1u$3",
    "automatic": "true"
  },
  "Domain Lowercase V": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([V])(.*)$",
    "replacement": "$1v$3",
    "automatic": "true"
  },
  "Domain Lowercase W": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([W])(.*)$",
    "replacement": "$1w$3",
    "automatic": "true"
  },
  "Domain Lowercase X": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([X])(.*)$",
    "replacement": "$1x$3",
    "automatic": "true"
  },
  "Domain Lowercase Y": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([Y])(.*)$",
    "replacement": "$1y$3",
    "automatic": "true"
  },
  "Domain Lowercase Z": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([Z])(.*)$",
    "replacement": "$1z$3",
    "automatic": "true"
  },
  "Domain made lowercase": {
    "regex": "^makedomainlowercase(.*)$",
    "replacement": "$1",
    "automatic": "true"
  }

```

This hack of serial regex replacements is very powerful and has the potential to solve a few issues or feature requests, but as you can see it is very verbose and potentially leaves a lot of "Applied" entries in the UI.

<details>

<summary>Screenshots</summary>

![Image](https://github.com/user-attachments/assets/fb7937db-7a6e-4b5f-8ddd-29b6c80b2af3)
![Image](https://github.com/user-attachments/assets/c2bef3c6-acf1-46b0-a1f5-b42993681ea5)
</details>

Not sure if/how it can be improved. Maybe adding an option "hideApplied" in the json to not show intermediary steps

```
  "Domain Lowercase A": {
    "regex": "^(makedomainlowercase.*?:\/\/[^\/?#]*)([A])(.*)$",
    "replacement": "$1a$3",
    "automatic": "true",
    "hideApplied": "true"
  },
  ...
```
But this only solves the UI part.

**Comment by TrianguloY** [comment]
That's...clever. Totally a hax, but hey it works!

I could add the hideApplied, but I would prefer to keep it explicit and instead find a better solution for this user-case.
